luma_core/src/ios.rs

```diff
@@ -0,0 +1,225 @@
+
+/// How to open a file.
+#[repr(i32)]
+pub enum Mode {
+    /// With no read/write access.
+    None = 0,
+
+    /// With read only access.
+    Read = 1,
+
+    /// With write only access.
+    Write = 2,
+
+    /// With read/write access.
+    ReadWrite = 3,
+}
+
+/// This is copied from std::io::SeekFrom, with the inner types changed to accomodate IOS.
+#[derive(Copy, PartialEq, Eq, Clone, Debug)]
+pub enum SeekFrom {
+    /// From the start of the file.
+    Start(i32),
+
+    /// From the current position.
+    Current(i32),
+
+    /// From the end of the file.
+    End(i32),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+#[repr(u32)]
+enum Command {
+    Open = 1,
+    Close = 2,
+    Read = 3,
+    Write = 4,
+    Seek = 5,
+    Ioctl = 6,
+    Ioctlv = 7,
+    Async = 8,
+}
+
+#[derive(Debug, Clone)]
+#[repr(C, align(32))]
+struct Ipc {
+    command: Command,
+    ret: i32,
+    fd: RawFd,
+    args: [i32; 5],
+}
+
+impl Ipc {
+    #[inline]
+    fn new(command: Command, args: [i32; 5]) -> Box<Self> {
+        Box::new(Ipc {
+            command,
+            ret: 0,
+            fd: -1,
+            args,
+        })
+    }
+
+    #[inline]
+    fn with_fd(command: Command, fd: i32, args: [i32; 5]) -> Box<Self> {
+        Box::new(Ipc {
+            command,
+            ret: 0,
+            fd,
+            args,
+        })
+    }
+
+    #[inline(never)]
+    fn send(self: Box<Self>) -> Box<Self> {
+        let ptr = Box::into_raw(self);
+
+        // Flush the IPC data from its cache line, so that the Starlet will see it.
+        unsafe { DCFlushRange(ptr as *const _, core::mem::size_of::<Self>() as u32) };
+
+        // Pass the pointer to IOS.
+        write32(HW_IPC_PPCMSG, ptr as u32 & 0x1fff_ffff);
+
+        // Signal to IOS we sent a command.
+        let ppcctrl = read32(HW_IPC_PPCCTRL);
+        if ppcctrl & 2 == 2 {
+            // TODO: Find out why Dolphin signals a command has been acknowledged already on boot.
+            write32(HW_IPC_PPCCTRL, 3);
+        } else {
+            write32(HW_IPC_PPCCTRL, 1);
+        }
+
+        // Busy loop until the Starlet acknowledges our command.
+        loop {
+            let ppcctrl = read32(HW_IPC_PPCCTRL);
+            if ppcctrl & 2 == 2 {
+                // Our command got acknowledged!
+                write32(HW_IPC_PPCCTRL, 2);
+                break;
+            }
+        }
+
+        // Busy loop until the Starlet replies to our command.
+        //
+        // TODO: provide an async API to avoid having to do that, for queries which are quite long.
+        loop {
+            let ppcctrl = read32(HW_IPC_PPCCTRL);
+            if ppcctrl & 4 == 4 {
+                // We got a reply!
+                break;
+            }
+        }
+
+        // Read the reply from IOS.
+        let armmsg = read32(HW_IPC_ARMMSG);
+        let command = unsafe { Box::from_raw((armmsg | 0x8000_0000) as *mut Ipc) };
+        assert_eq!(command.command, Command::Async);
```

Comment:
I don't think `Command::Async` is the right name for this. Maybe change `Async` to `Reply` ?